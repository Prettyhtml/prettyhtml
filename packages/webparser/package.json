{
  "name": "@starptech/webparser",
  "version": "0.8.5",
  "description": "Optimized html parser for HTML5 Web Components.",
  "main": "./dist/public_api.js",
  "scripts": {
    "build-ts": "tsc",
    "watch-ts": "tsc -w",
    "test": "jest --forceExit --detectOpenHandles",
    "test-coverage": "npm run test -- --coverage",
    "tslint": "tslint -c tslint.json -p tsconfig.json",
    "prepublish": "npm run build-ts"
  },
  "files": [
    "dist"
  ],
  "keywords": [
    "formatter",
    "webparser",
    "parser",
    "htmlparser",
    "html5"
  ],
  "author": "Dustin Deus <deusdustin@gmail.com> (http://www.starptech.de)",
  "license": "MIT",
  "devDependencies": {
    "jest": "^23.6.0",
    "ts-jest": "^23.10.4",
    "ts-node": "^7.0.1",
    "tslint": "^5.11.0",
    "tslint-eslint-rules": "^5.4.0",
    "typescript": "^3.1.1",
    "vrsource-tslint-rules": "^5.8.2"
  },
  "publishConfig": {
    "access": "public"
  }
}
