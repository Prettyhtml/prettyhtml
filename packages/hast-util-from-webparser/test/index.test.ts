import { HtmlParser } from '@starptech/webparser'
import fromWebParser from '../index'

function getParserOpt() {
  return {
    ignoreFirstLf: false,
    decodeEntities: false,
    selfClosingCustomElements: true,
    selfClosingElements: true
  }
}

test('Attributes', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<p id="foo" class="bar baz" data-qux="quux"></p>
  <p data-123="456"></p>
  <img alt hidden>
  <a download></a>
  <a download=example.mp3></a>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Element void', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `
  <img>
  <hr>
  <br>
  <meta>
  <area>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Simple', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(`<div><p>foo</p></div>`, 'TestComp')

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('SVG', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<div>
  <svg width="230" height="120" viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <circle cx="60"  cy="60" r="50" fill="red"/>
    <circle cx="170" cy="60" r="50" fill="green"/>
  </svg>
</div>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('SVG - should preserve explicit svg namespace', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<div>
  <svg:svg width="230" height="120" viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <svg:circle cx="60"  cy="60" r="50" fill="red"/>
    <svg:circle cx="170" cy="60" r="50" fill="green"/>
  </svg:svg>
</div>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Template', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<template id="text">!</template>
    <TEMPLATE id="html"><strong>importance</strong> and <em>emphasis</em>.</TEMPLATE>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Comment', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(`<!-- foo -->`, 'TestComp')

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Attributes with colon or @ as prefix', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(`<div :type="" @foo="bar"></div>`, 'TestComp')

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Attributes with colon, binding syntax and namespace', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<div id="app"><prism-editor :xlink:href="'myHref'" :foo="dddd" xmlns:xlink="http://www.w3.org/1999/xlink" /></div>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Case sensitive attributes', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(`<div ASYNC></div>`, 'TestComp')

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Doctype', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(`<!DOCTYPE html>`, 'TestComp')

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Doctype nameless', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(`<!DOCTYPE>`, 'TestComp')

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Doctype with html skeleton', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<!DOCTYPE><html><head></head><body>foo</body></html>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Attributes with namespace', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<svg xmlns:xlink="http://www.w3.org/1999/xlink"></svg>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})

test('Gaps detection - should set gapAfter data on elements followed by empty lines', () => {
  const parser = new HtmlParser(getParserOpt())
  const result = parser.parse(
    `<div></div>


    <img>

    <p>
      Hello

      <em>There</em>
    </p>
    <div></div>`,
    'TestComp'
  )

  expect(result.errors.length).toBe(0)

  const node = fromWebParser(result.rootNodes, {})

  expect(node).toMatchSnapshot()
})
