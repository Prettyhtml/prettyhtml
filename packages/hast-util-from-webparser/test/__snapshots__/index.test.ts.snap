// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attributes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 48,
          "line": 1,
          "offset": 48,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {
        "className": Array [
          "bar",
          "baz",
        ],
        "dataQux": "quux",
        "id": "foo",
      },
      "tagName": "p",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 2,
          "offset": 51,
        },
        "start": Object {
          "column": 48,
          "line": 1,
          "offset": 48,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 24,
          "line": 2,
          "offset": 73,
        },
        "start": Object {
          "column": 2,
          "line": 2,
          "offset": 51,
        },
      },
      "properties": Object {
        "data123": "456",
      },
      "tagName": "p",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 3,
          "offset": 76,
        },
        "start": Object {
          "column": 24,
          "line": 2,
          "offset": 73,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "properties": Object {
        "alt": "",
        "hidden": true,
      },
      "tagName": "img",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 4,
          "offset": 95,
        },
        "start": Object {
          "column": 18,
          "line": 3,
          "offset": 92,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 18,
          "line": 4,
          "offset": 111,
        },
        "start": Object {
          "column": 2,
          "line": 4,
          "offset": 95,
        },
      },
      "properties": Object {
        "download": true,
      },
      "tagName": "a",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 5,
          "offset": 114,
        },
        "start": Object {
          "column": 18,
          "line": 4,
          "offset": 111,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 30,
          "line": 5,
          "offset": 142,
        },
        "start": Object {
          "column": 2,
          "line": 5,
          "offset": 114,
        },
      },
      "properties": Object {
        "download": "example.mp3",
      },
      "tagName": "a",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Attributes with colon or @ as prefix 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 31,
          "line": 1,
          "offset": 31,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {
        ":type": "",
        "@foo": "bar",
      },
      "tagName": "div",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Attributes with colon, binding syntax and namespace 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "selfClosing": true,
          },
          "position": Object {
            "end": Object {
              "column": 108,
              "line": 1,
              "offset": 108,
            },
            "start": Object {
              "column": 14,
              "line": 1,
              "offset": 14,
            },
          },
          "properties": Object {
            ":foo": "dddd",
            ":xlink:href": "'myHref'",
            "xmlnsXLink": "http://www.w3.org/1999/xlink",
          },
          "tagName": "prism-editor",
          "type": "element",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 114,
          "line": 1,
          "offset": 114,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {
        "id": "app",
      },
      "tagName": "div",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Attributes with namespace 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 54,
          "line": 1,
          "offset": 54,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {
        "xmlnsXLink": "http://www.w3.org/1999/xlink",
      },
      "tagName": "svg",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Case sensitive attributes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 17,
          "line": 1,
          "offset": 17,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {
        "async": true,
      },
      "tagName": "div",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Comment 1`] = `
Object {
  "children": Array [
    Object {
      "position": Object {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 4,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "comment",
      "value": " foo ",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Doctype 1`] = `
Object {
  "children": Array [
    Object {
      "name": "html",
      "position": Object {
        "end": Object {
          "column": 15,
          "line": 1,
          "offset": 15,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "public": null,
      "system": null,
      "type": "doctype",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Doctype nameless 1`] = `
Object {
  "children": Array [
    Object {
      "name": "html",
      "position": Object {
        "end": Object {
          "column": 10,
          "line": 1,
          "offset": 10,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "public": null,
      "system": null,
      "type": "doctype",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Doctype with html skeleton 1`] = `
Object {
  "children": Array [
    Object {
      "name": "html",
      "position": Object {
        "end": Object {
          "column": 10,
          "line": 1,
          "offset": 10,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "public": null,
      "system": null,
      "type": "doctype",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 29,
              "line": 1,
              "offset": 29,
            },
            "start": Object {
              "column": 16,
              "line": 1,
              "offset": 16,
            },
          },
          "properties": Object {},
          "tagName": "head",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 38,
                  "line": 1,
                  "offset": 38,
                },
                "start": Object {
                  "column": 35,
                  "line": 1,
                  "offset": 35,
                },
              },
              "type": "text",
              "value": "foo",
            },
          ],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 45,
              "line": 1,
              "offset": 45,
            },
            "start": Object {
              "column": 29,
              "line": 1,
              "offset": 29,
            },
          },
          "properties": Object {},
          "tagName": "body",
          "type": "element",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 52,
          "line": 1,
          "offset": 52,
        },
        "start": Object {
          "column": 10,
          "line": 1,
          "offset": 10,
        },
      },
      "properties": Object {},
      "tagName": "html",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Element void 1`] = `
Object {
  "children": Array [
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 2,
          "offset": 3,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "properties": Object {},
      "tagName": "img",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 3,
          "offset": 11,
        },
        "start": Object {
          "column": 7,
          "line": 2,
          "offset": 8,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "properties": Object {},
      "tagName": "hr",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 4,
          "offset": 18,
        },
        "start": Object {
          "column": 6,
          "line": 3,
          "offset": 15,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "properties": Object {},
      "tagName": "br",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 5,
          "offset": 25,
        },
        "start": Object {
          "column": 6,
          "line": 4,
          "offset": 22,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "properties": Object {},
      "tagName": "meta",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 2,
          "line": 6,
          "offset": 34,
        },
        "start": Object {
          "column": 8,
          "line": 5,
          "offset": 31,
        },
      },
      "type": "text",
      "value": "
  ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "properties": Object {},
      "tagName": "area",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Gaps detection - should set gapAfter data on elements followed by empty lines 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "data": Object {
        "gapAfter": true,
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 11,
          "line": 1,
          "offset": 11,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "div",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 4,
          "line": 4,
          "offset": 18,
        },
        "start": Object {
          "column": 11,
          "line": 1,
          "offset": 11,
        },
      },
      "type": "text",
      "value": "


    ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "gapAfter": true,
        "selfClosing": false,
      },
      "properties": Object {},
      "tagName": "img",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 4,
          "line": 6,
          "offset": 29,
        },
        "start": Object {
          "column": 9,
          "line": 4,
          "offset": 23,
        },
      },
      "type": "text",
      "value": "

    ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 6,
              "line": 9,
              "offset": 52,
            },
            "start": Object {
              "column": 7,
              "line": 6,
              "offset": 32,
            },
          },
          "type": "text",
          "value": "
      Hello

      ",
        },
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 15,
                  "line": 9,
                  "offset": 61,
                },
                "start": Object {
                  "column": 10,
                  "line": 9,
                  "offset": 56,
                },
              },
              "type": "text",
              "value": "There",
            },
          ],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 20,
              "line": 9,
              "offset": 66,
            },
            "start": Object {
              "column": 6,
              "line": 9,
              "offset": 52,
            },
          },
          "properties": Object {},
          "tagName": "em",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 4,
              "line": 10,
              "offset": 71,
            },
            "start": Object {
              "column": 20,
              "line": 9,
              "offset": 66,
            },
          },
          "type": "text",
          "value": "
    ",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 8,
          "line": 10,
          "offset": 75,
        },
        "start": Object {
          "column": 4,
          "line": 6,
          "offset": 29,
        },
      },
      "properties": Object {},
      "tagName": "p",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 4,
          "line": 11,
          "offset": 80,
        },
        "start": Object {
          "column": 8,
          "line": 10,
          "offset": 75,
        },
      },
      "type": "text",
      "value": "
    ",
    },
    Object {
      "children": Array [],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 15,
          "line": 11,
          "offset": 91,
        },
        "start": Object {
          "column": 4,
          "line": 11,
          "offset": 80,
        },
      },
      "properties": Object {},
      "tagName": "div",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`SVG - should preserve explicit svg namespace 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 2,
              "line": 2,
              "offset": 8,
            },
            "start": Object {
              "column": 5,
              "line": 1,
              "offset": 5,
            },
          },
          "type": "text",
          "value": "
  ",
        },
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 4,
                  "line": 3,
                  "offset": 147,
                },
                "start": Object {
                  "column": 136,
                  "line": 2,
                  "offset": 142,
                },
              },
              "type": "text",
              "value": "
    ",
            },
            Object {
              "children": Array [],
              "data": Object {
                "selfClosing": true,
              },
              "position": Object {
                "end": Object {
                  "column": 52,
                  "line": 3,
                  "offset": 195,
                },
                "start": Object {
                  "column": 4,
                  "line": 3,
                  "offset": 147,
                },
              },
              "properties": Object {
                "cx": "60",
                "cy": "60",
                "fill": "red",
                "r": "50",
              },
              "tagName": "svg:circle",
              "type": "element",
            },
            Object {
              "position": Object {
                "end": Object {
                  "column": 4,
                  "line": 4,
                  "offset": 200,
                },
                "start": Object {
                  "column": 52,
                  "line": 4,
                  "offset": 195,
                },
              },
              "type": "text",
              "value": "
    ",
            },
            Object {
              "children": Array [],
              "data": Object {
                "selfClosing": true,
              },
              "position": Object {
                "end": Object {
                  "column": 54,
                  "line": 4,
                  "offset": 250,
                },
                "start": Object {
                  "column": 4,
                  "line": 4,
                  "offset": 200,
                },
              },
              "properties": Object {
                "cx": "170",
                "cy": "60",
                "fill": "green",
                "r": "50",
              },
              "tagName": "svg:circle",
              "type": "element",
            },
            Object {
              "position": Object {
                "end": Object {
                  "column": 2,
                  "line": 5,
                  "offset": 253,
                },
                "start": Object {
                  "column": 54,
                  "line": 5,
                  "offset": 250,
                },
              },
              "type": "text",
              "value": "
  ",
            },
          ],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 12,
              "line": 5,
              "offset": 263,
            },
            "start": Object {
              "column": 2,
              "line": 2,
              "offset": 8,
            },
          },
          "properties": Object {
            "height": "120",
            "viewBox": "0 0 200 200",
            "width": "230",
            "xmlns": "http://www.w3.org/2000/svg",
            "xmlnsXLink": "http://www.w3.org/1999/xlink",
          },
          "tagName": "svg:svg",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 0,
              "line": 6,
              "offset": 264,
            },
            "start": Object {
              "column": 12,
              "line": 5,
              "offset": 263,
            },
          },
          "type": "text",
          "value": "
",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 6,
          "line": 6,
          "offset": 270,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "div",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`SVG 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 2,
              "line": 2,
              "offset": 8,
            },
            "start": Object {
              "column": 5,
              "line": 1,
              "offset": 5,
            },
          },
          "type": "text",
          "value": "
  ",
        },
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 4,
                  "line": 3,
                  "offset": 143,
                },
                "start": Object {
                  "column": 132,
                  "line": 2,
                  "offset": 138,
                },
              },
              "type": "text",
              "value": "
    ",
            },
            Object {
              "children": Array [],
              "data": Object {
                "selfClosing": true,
              },
              "position": Object {
                "end": Object {
                  "column": 48,
                  "line": 3,
                  "offset": 187,
                },
                "start": Object {
                  "column": 4,
                  "line": 3,
                  "offset": 143,
                },
              },
              "properties": Object {
                "cx": "60",
                "cy": "60",
                "fill": "red",
                "r": "50",
              },
              "tagName": "circle",
              "type": "element",
            },
            Object {
              "position": Object {
                "end": Object {
                  "column": 4,
                  "line": 4,
                  "offset": 192,
                },
                "start": Object {
                  "column": 48,
                  "line": 4,
                  "offset": 187,
                },
              },
              "type": "text",
              "value": "
    ",
            },
            Object {
              "children": Array [],
              "data": Object {
                "selfClosing": true,
              },
              "position": Object {
                "end": Object {
                  "column": 50,
                  "line": 4,
                  "offset": 238,
                },
                "start": Object {
                  "column": 4,
                  "line": 4,
                  "offset": 192,
                },
              },
              "properties": Object {
                "cx": "170",
                "cy": "60",
                "fill": "green",
                "r": "50",
              },
              "tagName": "circle",
              "type": "element",
            },
            Object {
              "position": Object {
                "end": Object {
                  "column": 2,
                  "line": 5,
                  "offset": 241,
                },
                "start": Object {
                  "column": 50,
                  "line": 5,
                  "offset": 238,
                },
              },
              "type": "text",
              "value": "
  ",
            },
          ],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 8,
              "line": 5,
              "offset": 247,
            },
            "start": Object {
              "column": 2,
              "line": 2,
              "offset": 8,
            },
          },
          "properties": Object {
            "height": "120",
            "viewBox": "0 0 200 200",
            "width": "230",
            "xmlns": "http://www.w3.org/2000/svg",
            "xmlnsXLink": "http://www.w3.org/1999/xlink",
          },
          "tagName": "svg",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 0,
              "line": 6,
              "offset": 248,
            },
            "start": Object {
              "column": 8,
              "line": 5,
              "offset": 247,
            },
          },
          "type": "text",
          "value": "
",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 6,
          "line": 6,
          "offset": 254,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "div",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Simple 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 11,
                  "line": 1,
                  "offset": 11,
                },
                "start": Object {
                  "column": 8,
                  "line": 1,
                  "offset": 8,
                },
              },
              "type": "text",
              "value": "foo",
            },
          ],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 15,
              "line": 1,
              "offset": 15,
            },
            "start": Object {
              "column": 5,
              "line": 1,
              "offset": 5,
            },
          },
          "properties": Object {},
          "tagName": "p",
          "type": "element",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 21,
          "line": 1,
          "offset": 21,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "div",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;

exports[`Template 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 21,
              "line": 1,
              "offset": 21,
            },
            "start": Object {
              "column": 20,
              "line": 1,
              "offset": 20,
            },
          },
          "type": "text",
          "value": "!",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 32,
          "line": 1,
          "offset": 32,
        },
        "start": Object {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {
        "id": "text",
      },
      "tagName": "template",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 4,
          "line": 2,
          "offset": 37,
        },
        "start": Object {
          "column": 32,
          "line": 1,
          "offset": 32,
        },
      },
      "type": "text",
      "value": "
    ",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 42,
                  "line": 2,
                  "offset": 75,
                },
                "start": Object {
                  "column": 32,
                  "line": 2,
                  "offset": 65,
                },
              },
              "type": "text",
              "value": "importance",
            },
          ],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 51,
              "line": 2,
              "offset": 84,
            },
            "start": Object {
              "column": 24,
              "line": 2,
              "offset": 57,
            },
          },
          "properties": Object {},
          "tagName": "strong",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 56,
              "line": 2,
              "offset": 89,
            },
            "start": Object {
              "column": 51,
              "line": 2,
              "offset": 84,
            },
          },
          "type": "text",
          "value": " and ",
        },
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 68,
                  "line": 2,
                  "offset": 101,
                },
                "start": Object {
                  "column": 60,
                  "line": 2,
                  "offset": 93,
                },
              },
              "type": "text",
              "value": "emphasis",
            },
          ],
          "data": Object {
            "selfClosing": false,
          },
          "position": Object {
            "end": Object {
              "column": 73,
              "line": 2,
              "offset": 106,
            },
            "start": Object {
              "column": 56,
              "line": 2,
              "offset": 89,
            },
          },
          "properties": Object {},
          "tagName": "em",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 74,
              "line": 2,
              "offset": 107,
            },
            "start": Object {
              "column": 73,
              "line": 2,
              "offset": 106,
            },
          },
          "type": "text",
          "value": ".",
        },
      ],
      "data": Object {
        "selfClosing": false,
      },
      "position": Object {
        "end": Object {
          "column": 85,
          "line": 2,
          "offset": 118,
        },
        "start": Object {
          "column": 4,
          "line": 2,
          "offset": 37,
        },
      },
      "properties": Object {
        "id": "html",
      },
      "tagName": "TEMPLATE",
      "type": "element",
    },
  ],
  "data": Object {
    "selfClosing": false,
  },
  "type": "root",
}
`;
